pkg/operator/controller/dns_status.go

```diff
@@ -92,34 +166,114 @@ func computeDNSDegradedCondition(oldCondition *operatorv1.OperatorCondition, clu
 		maxUnavailable, intstrErr := intstr.GetScaledValueFromIntOrPercent(&maxUnavailableIntStr, int(want), true)
 		switch {
 		case want == 0:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "NoDNSPodsDesired")
-			messages = append(messages, "No DNS pods are desired; this could mean all nodes are tainted or unschedulable.")
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{Type: DNSNoDNSPodsDesired, Reason: DNSNoDNSPodsDesired, Status: operatorv1.ConditionTrue, Message: "No DNS pods are desired; this could mean all nodes are tainted or unschedulable."})
 		case have == 0:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "NoDNSPodsAvailable")
-			messages = append(messages, "No DNS pods are available.")
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{Type: DNSNoDNSPodsAvailable, Reason: DNSNoDNSPodsAvailable, Status: operatorv1.ConditionTrue, Message: "No DNS pods are available."})
 		case intstrErr != nil:
-			degradedReasons = append(degradedReasons, "InvalidDNSMaxUnavailable")
-			messages = append(messages, fmt.Sprintf("The DNS daemonset has an invalid MaxUnavailable value: %v", intstrErr))
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{Type: DNSInvalidDNSMaxUnavailable, Reason: DNSInvalidDNSMaxUnavailable, Status: operatorv1.ConditionTrue, Message: fmt.Sprintf("The DNS daemonset has an invalid MaxUnavailable value: %v", intstrErr)})
 		case int(numberUnavailable) > maxUnavailable:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "MaxUnavailableDNSPodsExceeded")
-			messages = append(messages, fmt.Sprintf("Too many DNS pods are unavailable (%d > %d max unavailable).", numberUnavailable, maxUnavailable))
+			graceCondition = &operatorv1.OperatorCondition{Type: DNSMaxUnavailableDNSPodsExceeded, Reason: DNSMaxUnavailableDNSPodsExceeded, Status: operatorv1.ConditionTrue, Message: fmt.Sprintf("Too many DNS pods are unavailable (%d > %d max unavailable).", numberUnavailable, maxUnavailable)}
+		}
+	}
+
+	// If there are degraded conditions, there is no grace period, and this should not be requeued.
+	// Set the final degraded condition as true, and concatenate the reasons and messages
+	if len(degradedConditions) != 0 {
+		degradedCondition.Status = operatorv1.ConditionTrue
+		degradedConditions = append(degradedConditions, setDNSLastTransitionTime(degradedCondition, oldDegradedCondition))
+
+		return degradedConditions, nil
+	}
+
+	// If the graceCondition occurs, this should be requeued.
+	// Set the final degraded condition as false, add the grace condition, and set it up to requeue the reconcile.
+	// We check this last, because the other conditions are non-recoverable as degraded conditions.
+	if graceCondition.Status == operatorv1.ConditionTrue {
```

Comment:
Using `checkConditions` neatly abstracted the grace logic, so the caller didn't need to special-case individual grace conditions, compute the requeue time, or have strange edge cases like `requeueTime == 0`.  Also, using `checkConditions` made the DNS operator more consistent with the ingress operator.  Any reason to rework the logic not to use `checkConditions`?